instrumentor/controllers/sourceinstrumentation/common.go

```diff
@@ -46,52 +47,63 @@ func syncNamespaceWorkloads(
 			continue
 		}
 
-		objects := make([]client.Object, 0)
 		switch obj := workloadObjects.(type) {
 		case *v1.DeploymentList:
 			for _, dep := range obj.Items {
-				objects = append(objects, &dep)
+				workloadsToSync = append(workloadsToSync, k8sconsts.PodWorkload{
+					Name:      dep.GetName(),
+					Namespace: dep.GetNamespace(),
+					Kind:      k8sconsts.WorkloadKindDeployment,
+				})
 			}
 		case *v1.DaemonSetList:
 			for _, ds := range obj.Items {
-				objects = append(objects, &ds)
+				workloadsToSync = append(workloadsToSync, k8sconsts.PodWorkload{
+					Name:      ds.GetName(),
+					Namespace: ds.GetNamespace(),
+					Kind:      k8sconsts.WorkloadKindDaemonSet,
+				})
 			}
 		case *v1.StatefulSetList:
 			for _, ss := range obj.Items {
-				objects = append(objects, &ss)
+				workloadsToSync = append(workloadsToSync, k8sconsts.PodWorkload{
+					Name:      ss.GetName(),
+					Namespace: ss.GetNamespace(),
+					Kind:      k8sconsts.WorkloadKindStatefulSet,
+				})
 			}
 		}
+	}
 
-		for _, obj := range objects {
-			workload := workload.ClientObjectFromWorkloadKind(kind)
-			err := k8sClient.Get(ctx, client.ObjectKeyFromObject(obj), workload)
-			if client.IgnoreNotFound(err) != nil {
-				return collectiveRes, err
-			}
-			res, err := syncWorkload(ctx, k8sClient, runtimeScheme, obj)
-			if err != nil {
-				errs = errors.Join(errs, err)
-			}
-			if !res.IsZero() {
-				collectiveRes = res
-			}
+	for _, pw := range workloadsToSync {
+		res, err := syncWorkload(ctx, k8sClient, runtimeScheme, pw)
+		if err != nil {
+			errs = errors.Join(errs, err)
+		}
+		if !res.IsZero() {
+			collectiveRes = res
 		}
 	}
+
 	return collectiveRes, errs
 }
 
 // syncWorkload checks if the given client.Object is instrumented by a Source.
 // If not, it will attempt to delete any InstrumentationConfig for the Object.
 // If it is instrumented, it will attempt to create an InstrumentationConfig if one does not exist,
 // or update the existing InstrumentationConfig if necessary.
-func syncWorkload(ctx context.Context, k8sClient client.Client, scheme *runtime.Scheme, obj client.Object) (ctrl.Result, error) {
+func syncWorkload(ctx context.Context, k8sClient client.Client, scheme *runtime.Scheme, pw k8sconsts.PodWorkload) (ctrl.Result, error) {
 	logger := log.FromContext(ctx)
 
-	pw := k8sconsts.PodWorkload{
-		Name:      obj.GetName(),
-		Namespace: obj.GetNamespace(),
-		Kind:      k8sconsts.WorkloadKind(obj.GetObjectKind().GroupVersionKind().Kind),
+	obj := workload.ClientObjectFromWorkloadKind(pw.Kind)
```

Comment:
@tamirdavid1  this call now happens once, it used to be invoked be each caller to `syncWorkload` (3 times, one for each caller)